### PR TITLE
QueryableByName supoort non ident column name

### PR DESCRIPTION
**Pull request by hi-rustin**
Try to close https://github.com/diesel-rs/diesel/issues/2747

close https://github.com/diesel-rs/diesel/pull/2772

close https://github.com/diesel-rs/diesel/pull/2770

### CODE REVIEW

**Comment by hi-rustin** [comment]
This PR is in a very rough stage.

/hold

- [ ] better name
- [ ] better tests

File: diesel_derives/src/field.rs [hunk 1]
```diff
@@ -65,6 +65,23 @@ impl Field {
             })
     }
 
+    pub fn column_str_name(&self) -> String {
+        self.column_name_from_attribute
+            .as_ref()
+            .map(|m| {
+                m.str_value().unwrap_or_else(|e| {
+                    e.emit();
+                    "unknown_column".to_string()
```
Comment (weiznich) [inline diesel_derives/src/field.rs:74]: If its a named field we should use the actual name of the field here, similary that done in the existing `column_name` function.

File: diesel_derives/src/field.rs [hunk 2]
```diff
@@ -65,6 +65,23 @@ impl Field {
...
+                })
+            })
+            .unwrap_or_else(|| {
+                self.span
+                    .error("All fields of tuple structs must be annotated with `#[column_name]`")
+                    .emit();
+                "unknown_column".to_string()
```
Comment (weiznich) [inline diesel_derives/src/field.rs:81]: Same as above

File: diesel_derives/src/queryable_by_name.rs [hunk 3]
```diff
@@ -22,15 +22,25 @@ pub fn derive(item: syn::DeriveInput) -> Result<proc_macro2::TokenStream, Diagno
...
+                            {
+                                let field = diesel::row::NamedRow::get(row, stringify!(#name))?;
+                                <#deserialize_ty as Into<#field_ty>>::into(field)
+                            }
+                    ))
+                } else {
+                    let name = f.column_str_name();
+                    Ok(quote!(
+                       {
+                           let field = diesel::row::NamedRow::get(row, stringify!(#name))?;
```
Comment (weiznich) [inline diesel_derives/src/queryable_by_name.rs:39]: As name already is a string, this stringify is not needed anymore.

**Comment by hi-rustin** [comment]
/unhold

Ready for review.

**Comment by hi-rustin** [comment]
https://github.com/diesel-rs/diesel/pull/2764/checks?check_run_id=2534309735 failed because I used the new API, but since this CI uses minimized packages, it will fail. But if I don't change it, other CIs will fail because of clippy. So maybe the solution is to allow the use of deprecated API in this place? Or is there another way to solve the problem?

**Comment by weiznich** [comment]
In this specific case its fine to just bump the minimal version of url to some version that supports the new conversion.

**Comment by hi-rustin** [comment]
> In this specific case its fine to just bump the minimal version of url to some version that supports the new conversion.

It was implemented on February 22, but was not released until yesterday. So I upgraded to 2.2.2.

File: diesel_derives/src/queryable_by_name.rs [hunk 4]
```diff
@@ -22,15 +22,25 @@ pub fn derive(item: syn::DeriveInput) -> Result<proc_macro2::TokenStream, Diagno
...
+                    ))
+                } else {
+                    let name = f.column_name_str();
+                    Ok(quote!(
+                       {
+                           let field = diesel::row::NamedRow::get(row, #name)?;
+                           <#deserialize_ty as Into<#field_ty>>::into(field)
+                       }
+                    ))
+                }
```
Comment (weiznich) [inline diesel_derives/src/queryable_by_name.rs:43]: I don't think we need to check if there is a `table_name` attribute or not. If there is a `column_name` attribute we should just unconditionally use the string value from there:

Comment (hi-rustin) [reply diesel_derives/src/queryable_by_name.rs]: Makes sense. Fixed.

**Review by weiznich — APPROVED** [review]

**Pull request merged**